jetty-server/src/main/java/org/eclipse/jetty/server/Server.java

```diff
@@ -466,21 +464,20 @@ protected void doStop() throws Exception
 
         MultiException mex = new MultiException();
 
-        try
+        if (getStopTimeout() > 0)
         {
-            // list if graceful futures
-            List<Future<Void>> futures = new ArrayList<>();
-            // First shutdown the network connectors to stop accepting new connections
-            for (Connector connector : _connectors)
+            try
             {
-                futures.add(connector.shutdown());
+                long end = System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(getStopTimeout());
+                Graceful.shutdown(this).get(getStopTimeout(), TimeUnit.MILLISECONDS);
+                QueuedThreadPool qtp = getBean(QueuedThreadPool.class);
+                if (qtp != null)
+                    qtp.setStopTimeout(Math.min(1000L, TimeUnit.NANOSECONDS.toMillis(end - System.nanoTime())));
```

Comment:
This logic seems wrong to me, we can never set the stopTimeout on the ThreadPool to longer than 1000ms, and if the `Graceful.shutdown` takes a full stopTimeout we could get a negative value for the ThreadPool stopTimeout.

Reply:
oops